smaps/src/main/java/com/google/smaps/SearchAddress.java

```diff
@@ -0,0 +1,136 @@
+/**
+ * Copyright 2020 Google LLC
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      https://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.google.smaps;
+
+import com.google.common.collect.RangeMap;
+import com.google.gson.Gson;
+import java.io.IOException;
+import java.math.BigInteger;
+import java.util.ArrayList;
+import java.util.List;
+import javax.servlet.annotation.WebServlet;
+import javax.servlet.http.HttpServlet;
+import javax.servlet.http.HttpServletRequest;
+import javax.servlet.http.HttpServletResponse;
+
+/**
+ * Retrieves the address entered by the user on memory-map.html, calculates which region occupies
+ * that address, and sends that region's index in the regions list to memory-map.js.
+ */
+@WebServlet(name = "SearchAddress", value = "/searchaddress")
+public class SearchAddress extends HttpServlet {
+  // Stores the String representation of the hex address as the user entered it.
+  static String address;
+  // Stores the address as a BigInteger that will be used to find the region occupying it.
+  static BigInteger addressBigInt;
+  // Stores the error message to display on memory-map.html.
+  static String errorMessage;
+
+  @Override
+  public void doPost(HttpServletRequest request, HttpServletResponse response) throws IOException {
+    // Reset the error message to be blank.
+    errorMessage = "";
+
+    // Check whether the reset button was clicked or not.
+    String resetButton = request.getParameter("reset-address-btn");
+    if (resetButton != null) {
+      // Reset was clicked, so remove the address from the searchbox and set addressBigInt to null.
+      address = "";
+      addressBigInt = null;
+    } else {
+      // Get the address the user entered.
+      address = request.getParameter("address-input");
+
+      // Use this regular expression to check that the address is actually hexadecimal, so that
+      // there isn't a NumberFormatException. If there are characters in the address that aren't in
+      // a valid hex number, set addressBigInt to null, and set errorMessage to proper error
+      // message.
+      if (address.matches("^[0-9a-fA-F]+$")) {
```

Comment:
A potential follow-on
- Tolerate a leading 0x, 0xF000
- Tolerate a trailing h, e.g., F000h
- Tolerate underscores, treat F000_F000 as F000F000
- Similarly tolerate spaces

These are nice policing features for UIs that accept addresses. The latter parts (_ and space) come into play with various conventions for human-readable addresses.